src/tensor_network.jl

```diff
@@ -80,7 +61,72 @@ Base.show(io::IO, ::MIME"text/plain", tn::TensorNetwork) = print(io, "TensorNetw
 
 next_tensor_id!(tn::TensorNetwork) = begin tn.next_id += 1; return Symbol("t$(tn.next_id - 1)") end
 bonds(tn::TensorNetwork) = keys(tn.bond_map)
-tensor_data(tn::TensorNetwork, i::Symbol) = tensor_data(tn.tensor_map[i])
+disable_hyperindices!(tn) = begin disable_hyperindices!.(tn); nothing; end
+
+
+"""
+    hyperindices(tn::TensorNetwork, i::Symbol; global_hyperindices=true)
+
+Find groups of hyper indices for the given tensor. When global_hyperindices is set to true, then
+indices which are identified as hyperindices because of groups of hyperindices in conneted tensors
+in the network are also included.
+"""
+function hyperindices(tn::TensorNetwork, i::Symbol; global_hyperindices=true, all_indices=false)
+    if !global_hyperindices
+        return hyperindices(tn[i], all_indices=all_indices)
+    else
+        all_hyperindices = Vector{Vector{Index}}()
+        tensor_indices = copy(inds(tn[i]))
+        while length(tensor_indices) > 0
+            index = tensor_indices[1]
+            connected_indices = find_connected_indices(tn, index)
+            push!(all_hyperindices, connected_indices)
+            setdiff!(tensor_indices, connected_indices)
+        end
+        return filter(x -> length(x) >= (all_indices ? 1 : 2), all_hyperindices)
+    end
+end
+
+"""
+    tensor_data(tn::TensorNetwork, i::Symbol; consider_hyperindices=true, global_hyperindices=true)
+
+Retrieve the tensor data for the given tensor. If the consider_hyperindices flag is true then
+then the data is reshaped to take into account the local hyperindices of the tensor. If the global_hyperindices
+index is also true then groups of hyperindices related via hyperindices for other tensors in the network are
+also considered.
+"""
+function tensor_data(tn::TensorNetwork, i::Symbol; consider_hyperindices=true, global_hyperindices=true)
+    if consider_hyperindices && global_hyperindices
+        global_hi = hyperindices(tn, i; global_hyperindices=true)
+        # filter any indices not connected to tensor
+        global_hi = map(global_hi) do x
+            x = filter(y -> y ∈ inds(tn[i]), x)
+        end
+        global_hi = filter(x -> length(x) > 1, global_hi)
+
+        local_hi = hyperindices(tn, i; global_hyperindices=false)
+        missing_hi = false
+        for g in global_hi
+            if findfirst(x -> Set(x) == Set(g), local_hi) === nothing
+                missing_hi = true
+            end
+        end
+        if missing_hi
+            t = store(tn[i])
+            if length(local_hi) > 0
```

Comment:
Just wondering if this if statement and the one on line 120 are needed. If looks like `missing_hi` will never be true if either `local_hi` or `global_hi` are empty.

Reply:
This logic is pretty ugly. You are right that if `global_hi` is empty then `missing_hi` will always be false so can remove the if statement at 120. I can make this change and commit.